DirectProgramming/DPC++FPGA/include/streaming_qri.hpp

```diff
@@ -0,0 +1,303 @@
+#pragma once
+
+#include "utils.hpp"
+#include "unrolled_loop.hpp"
+
+
+/*
+  QRI (QR inversion) - Given two matrices Q and R from the QR decomposition
+  of a matrix A such that A=QR, this function computes the inverse of A.
+  - Input matrix Q (unitary/orthogonal)
+  - Input matrix R (upper triangular)
+  - Output matrix I, the inverse of A such that A=QR
+
+  Then input and output matrices are consumed/produced from/to pipes.
+*/
+template <typename T,        // The datatype for the computation
+          bool is_complex,   // True if T is ac_complex<T>
+          int rows,          // Number of rows in the input matrices
+          int columns,       // Number of columns in the input matrices
+          int raw_latency,   // Read after write latency (in iterations) of
+                             // the triangular loop of this function.
+                             // This value depends on the FPGA target, the
+                             // datatype, the target frequency, etc.
+                             // This value will have to be tuned for optimal
+                             // performance. Refer to the Triangular Loop
+                             // design pattern tutorial.
+                             // In general, find a high value for which the
+                             // compiler is able to achieve an II of 1 and
+                             // go down from there.
+          int matrix_count,  // Number of matrices to read from the input
+                             // pipes sequentially
+          int pipe_size,     // Number of elements read/write per pipe
+                             // operation
+          typename QIn,      // Q input pipe, receive a full column with each
+                             // read.
+          typename RIn,      // R input pipe. Receive one element per read.
+                             // Only upper-right elements of R are sent.
+                             // Sent in row order, starting with row 0.
+          typename IOut      // Inverse matrix output pipe.
+                             // The output is written column by column
+          >
+struct StreamingQRI {
+  void operator()() const {
+    // Functional limitations
+    static_assert(rows == columns,
+                  "only square matrices with rows==columns are supported");
+    static_assert((columns <= 512) && (columns >= 4),
+                  "only matrices of size 4x4 to 512x512 are supported");
```

Comment:
Why the upper limit of 512?

Reply:
Upper limit to what I have tested, but this limit can be removed as the implementation as no limits (as long as the HW can implement it)